File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
'''
최단거리 문제
bfs를 통해 해당 좌표에 도달 시 depth를 구한다.

캐릭터는 항상 (0,0) 목적지는 (n-1,m-1)

'''
from collections import deque
def solution(maps):
    n,m = len(maps), len(maps[0])
    #print(n,m)
    def bfs(start):
        visited = set()
        queue = deque()
        # 갈 곳을 queue에 넣는다.
        queue.append((0,0,1))
        visited.add((0,0))
        while queue:
            y,x,d = queue.popleft()
            if (y,x) == (n-1,m-1):
                return d
            for dy, dx in [(0,1),(1,0),(-1,0),(0,-1)]:
                yy = y + dy
                xx = x + dx
                if 0 <= yy < n and 0 <= xx < m and maps[yy][xx] == 1:
                    if (yy,xx) not in visited:
                        queue.append((yy,xx,d+1))
                        visited.add((yy,xx))
        return -1
            
        # queue에서 popleft한다.
        # 해당 지점 방문 여부 판단하여 방문 안했을 시에만 다음 갈 곳 queue에 넣기
    
    return bfs((0,0))
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public static void log(Object o){
        System.out.println(o);
    }
    public int bfs(int[][] map, int startY, int startX, int endY, int endX){
        boolean[][] visited = new boolean[endY+1][endX+1];
        Queue<int[]> queue = new LinkedList<>();
        int[] start = {startY, startX, 0};
        queue.offer(start);
        visited[startY][startX] = true;
        
        
        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            // log(cur.get(0) + " " +cur.get(1));
            if(cur[0] == endY && cur[1] == endX){
                return cur[2] + 1;
            }
            int[][] directions = {
                {1,0}, {0,1},{-1,0},{0,-1}
            };
            for(int[] d: directions){
                int dy = d[0], dx = d[1];
                int yy = cur[0] + dy, xx = cur[1] + dx;
                if( (0<= yy && yy <= endY) && (0<= xx && xx <= endX) && map[yy][xx] == 1){
                    if(!visited[yy][xx]){
                        queue.offer(new int[]{ yy, xx, cur[2]+1 });
                        visited[yy][xx] = true;
                    }
                    
                }
            }
        }
        return -1;
    }
    public int solution(int[][] maps) {
        //bfs
        int n = maps.length, m = maps[0].length;
        int result = bfs(maps, 0,0, n-1,m-1);
        log(result);
        return result;
    }
}